runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,162 @@
+  end
+end
+
+---@private
+local function modifiers_from_number(x, modifiers_table)
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+--- 
+--- This function can be configured with |vim.lsp.with()| with the following options for `config`
+--- 
+--- `on_token`: A function with signature `function(ctx, token)` that is called
+---             whenever a semantic token is received from the server from context `ctx`
+---             (see |lsp-handler| for the definition of `ctx`). This can be used for highlighting the tokens.
+---             `token` is a table:
+---
+--- <pre>
+---   {
+---         line       -- line number 0-based
+---         start_char -- start character 0-based
+---         length     -- length in characters of this token
+---         type       -- token type as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         modifiers  -- token modifier as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens 
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  if not ctx.bufnr or not ctx.client_id then
+    return
+  end
+  active_requests[ctx.bufnr] = false
+  if config and config.on_invalidate_range then
+    config.on_invalidate_range(ctx, 0, -1)
+  end
+  -- if tick has changed our response is outdated!
+  if err or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    semantic_tokens[ctx.bufnr][ctx.client_id] = {}
+    return
+  end
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+
+  local tokens = {}
+  local line, start_char = nil, 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local token = {
```

Comment:
Technically we could just leave the start_char as-is, and include the offset encoding. I'm not sure what the best thing to do is here, computing the true start_char involves fetching the line from the buffer so it could also be handled when the highlights are applied. We could just include offset_encoding as an additional table field here for now.

Reply:
Also thought about this, the bug mentioned https://github.com/theHamsta/nvim-semantic-tokens/blob/77f5b716ecb13e4c692d4ec442da75ba850d8441/lua/nvim-semantic-tokens.lua#L58 is probably outside of this PR where I just assume bytes.

Reply:
@zbirenbaum  I only tested this PR with clangd using this plugin https://github.com/theHamsta/nvim-semantic-tokens . It would be important to validate this PR also with other LSP implementations like zls or sumneko_lua. rust-analyzer does not use standard-conformant semantic tokens. You could test manually and dump the responses from the LSP servers on a simple file for tests.

Reply:
@zbirenbaum also just did a few updates after following the review comments here. I will have earliest time for this on weekend and last weekend I was working on nvim-treesitter instead.

Reply:
FYI I used this with rust-analyzer during AOC and everything worked fine :)

Reply:
Really? I saw all the mapping to to highlighting in the typescript part of rust-analyzer an guessed that their implementation predated semantic tokens. But that's nice to hear!

Reply:
@theHamsta I planned to use pylance as I have it running in cli, but don't tell microsoft, not sure it gets much less standard than that. I was unaware sumneko supported semantic highlights but I will try it with that too. Thanks for the updates and info, i'll see what I can do.

Thanks for working on treesitteer by the way!

Reply:
It's not that it has to be tested on any configuration. But some kind of validation or polishing of rough edges would be good.

Reply:
I noticed pylance has a ``namespace`` tokenType that is missing here.

Reply:
That's not in the official specification so it would need to be added by a plugin, not here.

Reply:
> I noticed pylance has a `namespace` tokenType that is missing here. Never mind, that was what vscode sent to the server.

How are you seeing the communication between pylance and vscode? Pylance was strange with this PR, as it errors stating that no server supports semantic highlights on first request, but then on the next request for them functions perfectly fine. I'm curious if it's a result of some off spec implementation or an edge case of the code in this PR that needs handling.

Reply:
Well, I have the possibility to react to errors. If the error message is helpful that it sends (.i.e. not ready I could react to it. Maybe we can also check the capabilities dynamically

Reply:
> Well, I have the possibility to react to errors. If the error message is helpful that it sends (.i.e. not ready I could react to it. Maybe we can also check the capabilities dynamically

Here's the output I get from opening a lua file:
![2022-02-11-031807_929x444_scrot](https://user-images.githubusercontent.com/32016110/153558397-649676d4-818b-49f9-9cdb-b6d1a75236ee.png)
![2022-02-11-031841_1085x211_scrot](https://user-images.githubusercontent.com/32016110/153558398-01e71164-86ac-4d86-8279-fa08ee8224eb.png)

I am also getting this issue with clangd, although I know for certain I was not before. This may be a side effect of rebasing the changes onto an up to date neovim. While there weren't any conflicts in doing so, and this behavior was present prior for pylance, it was not (that I can recall) for clangd and lua.

Reply:
> > I noticed pylance has a `namespace` tokenType that is missing here. Never mind, that was what vscode sent to the server.
> 
> How are you seeing the communication between pylance and vscode? Pylance was strange with this PR, as it errors stating that no server supports semantic highlights on first request, but then on the next request for them functions perfectly fine. I'm curious if it's a result of some off spec implementation or an edge case of the code in this PR that needs handling.
> 
> I believe that the change to make pylance vscode-only occurred after semantic highlights were implemented in it, so anyone else who wants to test this could probably download an old version and use it without having to be forever labeled by microsoft's omniscient lawyers as a license breaker.
> 
> Edit: @theHamsta It is almost assuredly an edge case. Calling semantic_tokens_full() from autocmd causes the same issue with lua language server. It was 'fixable' by forcing the autocommand to wait to be registered until after the 'done' field was received in progress messages. Unfortunately, pylance did not have any real progress messages, and I don't love the idea of making functionality dependent on a different language server capability if there are other possibilities.
> 
> I can't really think of a 'good' way to do this, but simply suppressing the output by modifying the autocommand for your plugin to: `vim.cmd[[ autocmd BufEnter,CursorHold,InsertLeave <buffer> silent! lua require'vim.lsp.buf'.semantic_tokens_full() ]]` works with all LSPs I tested it with, inelegant as it may be (although what looks like an error flashes for a few ms on opening a buffer)

I did a wrapper around the server (ipc process) and logged the input and output. I'm also using the latest version with no issues. I'm enabling it inside the on_attach.
```lua
local on_attach = function(client, bufnr)
    if client.resolved_capabilities.semantic_tokens_full == true then
        vim.cmd [[autocmd BufEnter,CursorHold,InsertLeave <buffer> lua vim.lsp.buf.semantic_tokens_full()]]
    end
end
```

Reply:
> I noticed pylance has a `namespace` tokenType that is missing here. Never mind, that was what vscode sent to the server.

well spotted! the `namespace` token is also mentioned in the LSP specification. https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#semanticTokenTypes

Reply:
> I did a wrapper around the server (ipc process) and logged the input and output. I'm also using the latest version with no issues. I'm enabling it inside the on_attach.
> 
> ```lua
> local on_attach = function(client, bufnr)
>     if client.resolved_capabilities.semantic_tokens_full == true then
>         vim.cmd [[autocmd BufEnter,CursorHold,InsertLeave <buffer> lua vim.lsp.buf.semantic_tokens_full()]]
>     end
> end
> ```

Placing that inside of attach hook works perfectly with pylance, lua, and clangd. "== true" is a bit repetitive and can be simplified to:
 `if client.resolved_capabilities.semantic_tokens_full then ...`
but I would recommend putting the snippet you provided with that slight adjustment into in the official documentation for this PR.
`

Reply:
Another thing I noticed is that pylance's `textDocument/semanticTokens/full/delta` is just a copy of its `textDocument/semanticTokens/full`, all tokens are sent, clangd's `textDocument/semanticTokens/full/delta` works like in the lsp specs. Clangd's `textDocument/semanticTokens/range` is not implemented, but pylance's is. This might or might not help you test those features.